src/Umbraco.Web/Models/Mapping/ContentVariantMapper.cs

```diff
@@ -21,52 +21,110 @@ public ContentVariantMapper(ILocalizationService localizationService)
-                }).ToList();
+                    // This should not happen
+                    throw new InvalidOperationException("No languages or segments available");
+                }
 
-                for (int i = 0; i < langs.Count; i++)
-                {
-                    var x = langs[i];
-                    var variant = variants[i];
+                variants = languages
+                    .SelectMany(language => segments
+                        .Select(segment => CreateVariantDisplay(context, source, language, segment)))
+                    .ToList();
+            }
 
-                    variant.Language = x;
-                    variant.Name = source.GetCultureName(x.IsoCode);
-                }
+            return SortVariants(variants);
+        }
 
-                //Put the default language first in the list & then sort rest by a-z
-                var defaultLang = variants.SingleOrDefault(x => x.Language.IsDefault);
+        private IList<ContentVariantDisplay> SortVariants(IList<ContentVariantDisplay> variants)
+        {
+            if (variants == null || variants.Count <= 1)
+            {
+                return variants;
+            }
 
-                //Remove the default language from the list for now
-                variants.Remove(defaultLang);
+            // Default variant first, then order by language, segment.
+            return variants
+                .OrderBy(v => IsDefaultLanguage(v) ? 0 : 1)
+                .ThenBy(v => IsDefaultSegment(v) ? 0 : 1)
+                .ThenBy(v => v?.Language?.Name)
+                .ThenBy(v => v.Segment)
+                .ToList();
+        }
 
-                //Sort the remaining languages a-z
-                variants = variants.OrderBy(x => x.Language.Name).ToList();
+        private static bool IsDefaultSegment(ContentVariantDisplay variant)
+        {
+            return variant.Segment == null;
+        }
 
-                //Insert the default language as the first item
-                variants.Insert(0, defaultLang);
+        private static bool IsDefaultLanguage(ContentVariantDisplay variant)
+        {
+            return variant.Language == null || variant.Language.IsDefault;
+        }
 
-                return variants;
+        private IEnumerable<Language> GetLanguages(MapperContext context)
+        {
+            var allLanguages = _localizationService.GetAllLanguages().OrderBy(x => x.Id).ToList();
+            if (allLanguages.Count == 0)
+            {
+                // This should never happen
+                return Enumerable.Empty<Language>();
+            }
+            else
+            {
+                return context.MapEnumerable<ILanguage, Language>(allLanguages).ToList();
             }
-            return result;
+        }
+
+        private IEnumerable<string> GetSegments(IContent content)
+        {
+            return content.Properties.SelectMany(p => p.Values.Select(v => v.Segment)).Distinct();
```

Comment:
@PerplexDaniel  Just testing some stuff now, is it expected that this will return a `null` value as part of the result? or should `null` be filtered from here? I can't quite tell if that's expected or not, else this needs a WhereNotNull appended to it. I've merged and tested this in my temp branch and it works, i'll merge into dev/v8 once you let me know about this null.

Reply:
Yeah, segment `null` would be the "default" segment, sort of like the default language when a content type varies by language. In the above context the variants of a page are created based on the segments, and segment `null` should definitely be part of it as one of the variants will have segment `null` and another could have a segment `segment-b` or whatever alias is chosen. So when one segment is added the page would have at least 2 variants (one with segment `null` and another one with an actual segment alias).

I believe existing core code elsewhere will convert `null` to the empty string, and this is also how you can explicitly read the unsegmented property value in the front-end: `Model.Value("myProp", segment: "")`. Passing in `null` would fallback to the current variation context instead.

Reply:
Ok cool, i'll leave it as-is and add a note to it.